selfdrive/car/subaru/carstate.py

```diff
@@ -47,6 +50,16 @@ def update(self, cp, cp_cam, cp_body):
     ret.gearShifter = self.parse_gear_shifter(self.shifter_values.get(can_gear, None))
 
     ret.steeringAngleDeg = cp.vl["Steering_Torque"]["Steering_Angle"]
+
+    # have not found a steering rate message. calculate it manually
+    STEERING_FREQUENCY = 2
+    ret.steeringRateDeg = (ret.steeringAngleDeg - self.prev_angle) / (DT_CTRL * STEERING_FREQUENCY)
```

Comment:
have you looked for any bits that change before a steering fault occurs? might want to look up to half a second back. I have this script which I modify, for example change the comparison to a variable like `bit_to_find` and set that appropriately above: https://gist.github.com/sshane/a7965aefd6462186edc6152ccf203eed#file-find_torque_sign_bit-py-L63

Reply:
I've done some searching in cabana but I'll look at making a script

Reply:
I looked with a script and couldn't find anything. In techinfo, I can see that one ES cancel condition is "steering angle changed excessively", but in ssm4 I don't see a "steering rate" message on either the EPS or ES, so my assumption is that both the EPS and ES calculate these manually, ES typically cancels before the steer fault, and since we don't cancel, the EPS faults.